src/core/thread_pool.cpp

```diff
@@ -17,91 +30,91 @@ ThreadPool::ThreadPool(uint32_t threads_num)
         }
         for (uint32_t i = 0; i < m_nr_threads - 1; i++) {
             m_workers.push_back(new Worker([this, i]() {
-                while (!m_stop) {
-                    while (m_active) {
-                        //! if the thread should work
-                        if (m_workers[i]->work_flag) {
-                            // printf("thread %d work form %d to %d\n", i,
-                            //        i * m_task_per_thread,
-                            //        (i + 1) * m_task_per_thread);
-                            m_task(TaskId{
-                                    i * m_task_per_thread,
-                                    std::min((i + 1) * m_task_per_thread, m_nr_task),
-                                    i});
-                            // printf("thread %d finished\n", i);
-                            //! Flag worker is finished
-                            m_workers[i]->work_flag = false;
-                        }
-                        //! Wait next task coming
-                        std::this_thread::yield();
-                    }
-                    {
-                        std::unique_lock<std::mutex> lock(m_mutex);
-                        if (!m_stop && !m_active) {
-                            m_cv.wait(lock, [this] { return m_stop || m_active; });
+                {
+                    std::unique_lock<std::mutex> lk(m_mtx);
+                    m_cv.wait(lk);
+                }
+                auto& w = m_workers[i]->work_flag;
+                while (true) {
+                    for (int j = 0;; j++) {
+                        worker_status_t s = w.load(std::memory_order_consume);
+                        if (s == WRK_START) {
+                            break;
+                        } else if (s == WRK_STOP)
+                            return;
+                        if (s == WRK_IDLE) {
+                            std::unique_lock<std::mutex> lk(m_mtx);
+                            m_cv.wait(lk);
                         }
+                        if (j & 0x1F)
```

Comment:
and a state `else if (s == BUSY_WAIT){` here ? and if the `s` rename with `state` is better

Reply:
I try to reduce `if else` here. I found that branches matters in such a tight loop.

I mean possibly `0x1F` or `0xFF` consective atomic ops.